#!/bin/bash
#
# BayesAss3-SNPs: Modification of BayesAss 3.0.4 to allow handling of large SNP datasets (https://github.com/stevemussmann/BayesAss3-SNPs)
# BA3-SNPS-autotune: Python code for automatically tuning mixing parameters in BayesAss3 (https://github.com/stevemussmann/BA3-SNPS-autotune)
#
# Requirements
#  VCFtools: A set of tools written in Perl and C++ for working with VCF files, such as those generated by the 1000 Genomes Project (https://github.com/vcftools/vcftools)
#  Stacks: a pipeline for building loci from short-read sequences, such as those generated on the Illumina platform (https://catchenlab.life.illinois.edu/stacks/)
#  A collection of file format converters to prepare input for several popular phylogenetic and population genetics software packages (https://github.com/stevemussmann/file_converters)
#
# Thin VCF file
vcftools --vcf sailfish.pruned.vcf --thin 50000 --remove-indels --recode --recode-INFO-all --out sailfish.pruned.thin.vcf

# Convert VCF in Stacks structure file
${HOME}/software/stacks-2.60/populations -V sailfish.pruned.thin.vcf.recode.vcf -M popmap -O . -t 16 --structure

# convert stacks structure into BayesAss input file
${HOME}/software/file_converters/stacksStr2immanc.pl -o sailfish.immanc -p WCA,SWA,ECA,SWI,SEI,WCP -s sailfish.pruned.thin.vcf.recode.p.structure

# run the atotune BayesAss to find suitable parameters
python3 ${HOME}/software/BA3-SNPS-autotune-2.1.2/BA3-SNPS-autotune.py -i sailfish.immanc -l 12034 -b 100000 -g 1000000 -r 10

# Run three independendt runs of BayesAss3-SNPs
cat bayesass_commands.sh | parallel -j 3 --tmpdir ${HOME}/tmp

